subsys/net/l2/ethernet/ethernet.c

```diff
@@ -507,22 +507,48 @@ static struct net_buf *ethernet_fill_header(struct ethernet_context *ctx,
 					    struct net_pkt *pkt,
 					    uint32_t ptype)
 {
+	struct net_if *orig_iface = iface;
 	struct net_buf *hdr_frag;
 	struct net_eth_hdr *hdr;
-	size_t hdr_len = IS_ENABLED(CONFIG_NET_VLAN) ?
-			 sizeof(struct net_eth_vlan_hdr) :
-			 sizeof(struct net_eth_hdr);
+	size_t reserve_ll_header;
+	size_t hdr_len;
+	bool is_vlan;
 
-	hdr_frag = net_pkt_get_frag(pkt, hdr_len, NET_BUF_TIMEOUT);
-	if (!hdr_frag) {
-		return NULL;
+	is_vlan = IS_ENABLED(CONFIG_NET_VLAN) &&
+		net_eth_is_vlan_enabled(ctx, iface) &&
+		net_pkt_vlan_tag(pkt) != NET_VLAN_TAG_UNSPEC;
+	if (is_vlan) {
+		orig_iface = net_eth_get_vlan_iface(iface, net_pkt_vlan_tag(pkt));
 	}
 
-	if (IS_ENABLED(CONFIG_NET_VLAN) &&
-	    net_eth_is_vlan_enabled(ctx, iface) &&
-	    net_pkt_vlan_tag(pkt) != NET_VLAN_TAG_UNSPEC) {
+	reserve_ll_header = net_if_reserve_ll_header_size(orig_iface);
```

Comment:
Should we check length here?

```c
if (reserve_ll_header > CONFIG_NET_BUF_DATA_SIZE) {
	return NULL;
}
```

Reply:
There is already an assert for that in net_if.c `net_if_reserve_ll_header_size()` it should be enough.

Reply:
Better safe than sorry?

Reply:
The header size is suppose to be set by the Ethernet or VLAN code and with a proper value, so a component we are trusting. If someone else writes some garbage to it, we do assert. Isn't this pretty much the way of working in zephyr?